File: api/ORM/api/project_stats.py
from fastapi import APIRouter, Query, HTTPException
from typing import Optional
from project_stats_models import project_stats
from tortoise.expressions import Q
from pydantic import BaseModel

project_stats_api = APIRouter()

class projectstatsIn(BaseModel):
    id: Optional[int] = None
    project_name: str  # 保持必填
    work_area: str  # 保持必填
    bid_money: Optional[str] = None
    bid_time: Optional[str] = None
    finish_time: Optional[str] = None
    tunnel_length: Optional[str] = None
    tube_outer_diameter: Optional[str] = None
    through_waters: Optional[str] = None
    max_excavation_diameter: Optional[str] = None
    project_type: Optional[str] = None
    job_schedule: Optional[str] = None
    type: Optional[str] = None
    company_name: Optional[str] = None
    province: Optional[str] = None
    remark: Optional[str] = None
    pro_code: Optional[str] = None
    sub_code: Optional[str] = None
    through_assise: Optional[str] = None
    hydraulic_pressure: Optional[str] = None
    intension: Optional[str] = None
    duct_piece: Optional[str] = None
    zone_id: Optional[int] = None

class projectstatsupdatemodel(BaseModel):
    id: int  # ID是必填的，用于更新记录
    project_name: Optional[str] = None
    work_area: Optional[str] = None
    bid_money: Optional[str] = None
    bid_time: Optional[str] = None
    finish_time: Optional[str] = None
    tunnel_length: Optional[str] = None
    tube_outer_diameter: Optional[str] = None
    through_waters: Optional[str] = None
    max_excavation_diameter: Optional[str] = None
    project_type: Optional[str] = None
    job_schedule: Optional[str] = None
    type: Optional[str] = None
    company_name: Optional[str] = None
    province: Optional[str] = None
    remark: Optional[str] = None
    pro_code: Optional[str] = None
    sub_code: Optional[str] = None
    through_assise: Optional[str] = None
    hydraulic_pressure: Optional[str] = None
    intension: Optional[str] = None
    duct_piece: Optional[str] = None
    zone_id: Optional[int] = None

@project_stats_api.post("/project_stats")
async def add_project_stats(stats_in: projectstatsIn):
    try:
        new_stats = await project_stats.create(**stats_in.dict(exclude_unset=True))
        return {"status": 0, "msg": "添加成功", "data": {"id": new_stats.id}}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"创建项目统计失败: {str(e)}")

@project_stats_api.get("/project_stats_list")
async def get_project_stats_list(
    id: Optional[int] = Query(None, description="编号"),
    project_name: Optional[str] = Query(None, description="项目名称"),
    work_area: Optional[str] = Query(None, description="工区"),
    pro_code: Optional[str] = Query(None, description="项目编码"),
    page: int = Query(1, ge=1, description="页码"),
    per_page: int = Query(10, ge=1, le=100, description="每页数量")
):
    filters = Q()
    if id is not None:
        filters &= Q(id=id)
    if project_name:
        filters &= Q(project_name=project_name)
    if work_area:
        filters &= Q(work_area=work_area)
    if pro_code:
        filters &= Q(pro_code=pro_code)

    total = await project_stats.filter(filters).count()
    items = await project_stats.filter(filters)\
        .offset((page - 1) * per_page)\
        .limit(per_page)

    return {
        "status": 0,
        "msg": "ok",
        "data": {
            "items": [{
                "id": item.id,
                "project_name": item.project_name,
                "work_area": item.work_area,
                "bid_money": item.bid_money,
                "bid_time": item.bid_time,
                "finish_time": item.finish_time,
                "tunnel_length": item.tunnel_length,
                "tube_outer_diameter": item.tube_outer_diameter,
                "through_waters": item.through_waters,
                "max_excavation_diameter": item.max_excavation_diameter,
                "project_type": item.project_type,
                "job_schedule": item.job_schedule,
                "type": item.type,
                "company_name": item.company_name,
                "province": item.province,
                "remark": item.remark,
                "pro_code": item.pro_code,
                "sub_code": item.sub_code,
                "through_assise": item.through_assise,
                "hydraulic_pressure": item.hydraulic_pressure,
                "intension": item.intension,
                "duct_piece": item.duct_piece,
                "zone_id": item.zone_id
            } for item in items],
            "total": total,
            "page": page,
            "per_page": per_page
        }
    }

@project_stats_api.post("/update_project_stats")
async def update_project_stats(stats_data: projectstatsupdatemodel):
    try:
        stats = await project_stats.get_or_none(id=stats_data.id)
        if not stats:
            return {"status": 1, "msg": f"未找到ID为{stats_data.id}的记录"}
        
        await stats.update_from_dict(stats_data.dict(exclude_unset=True))
        await stats.save()
        return {"status": 0, "msg": "更新成功"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"更新失败: {str(e)}")

@project_stats_api.delete("/delete_project_stats/{id}")
async def delete_project_stats(id: int):
    try:
        stats = await project_stats.get_or_none(id=id)
        if not stats:
            return {"status": 1, "msg": f"未找到ID为{id}的记录"}
        
        await stats.delete()
        return {"status": 0, "msg": "删除成功"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"删除失败: {str(e)}")

File: pages/project_stats.js
(function () {
  const response = {
    "type": "page",
    "body": [
      {
        "type": "button",
        "label": "新增项目统计",
        "icon": "fa fa-plus",
        "level": "primary",
        "className": "m-b-sm",
        "actionType": "dialog",
        "dialog": {
          "title": "新增项目统计",
          "body": {
            "type": "form",
            "api": {
              "method": "post",
              "url": "/project_stats/project_stats",
              "messages": {
                "success": "添加成功",
                "failed": "添加失败"
              }
            },
            "body": [
              {
                "type": "input-text",
                "name": "project_name",
                "label": "项目名称",
                "required": true
              },
              {
                "type": "input-text",
                "name": "work_area",
                "label": "工区",
                "required": true
              },
              {
                "type": "input-text",
                "name": "bid_money",
                "label": "中标金额",
                "required": true
              },
              {
                "type": "input-text",
                "name": "bid_time",
                "label": "中标时间",
                "required": true
              },
              {
                "type": "input-text",
                "name": "finish_time",
                "label": "完工时间"
              },
              {
                "type": "input-text",
                "name": "tunnel_length",
                "label": "隧道长度"
              },
              {
                "type": "input-text",
                "name": "tube_outer_diameter",
                "label": "管片外径"
              },
              {
                "type": "input-text",
                "name": "through_waters",
                "label": "是否穿越水域"
              },
              {
                "type": "input-text",
                "name": "max_excavation_diameter",
                "label": "最大开挖直径"
              },
              {
                "type": "input-text",
                "name": "project_type",
                "label": "工程类型"
              },
              {
                "type": "input-text",
                "name": "job_schedule",
                "label": "工程进度"
              },
              {
                "type": "input-text",
                "name": "type",
                "label": "后构类型"
              },
              {
                "type": "input-text",
                "name": "company_name",
                "label": "单位名称"
              },
              {
                "type": "input-text",
                "name": "province",
                "label": "省份"
              },
              {
                "type": "input-text",
                "name": "remark",
                "label": "备注"
              },
              {
                "type": "input-text",
                "name": "pro_code",
                "label": "项目编码"
              },
              {
                "type": "input-text",
                "name": "sub_code",
                "label": "项目区间编码"
              },
              {
                "type": "input-text",
                "name": "through_assise",
                "label": "穿越地层"
              },
              {
                "type": "input-text",
                "name": "hydraulic_pressure",
                "label": "埋深水压"
              },
              {
                "type": "input-text",
                "name": "intension",
                "label": "衬砌强度"
              },
              {
                "type": "input-text",
                "name": "duct_piece",
                "label": "隧道管片"
              },
              {
                "type": "input-number",
                "name": "zone_id",
                "label": "项目区间id"
              },
              {
                "type": "input-number",
                "name": "shield_machine_id",
                "label": "盾构机ID"
              }
            ]
          }
        }
      },
      {
        "type": "crud",
        "syncLocation": false,
        "resizable": true,
        "columnsTogglable": true,
        "className": "table-db m-b-none",
        "tableClassName": "table-cell-nowrap",
        "api": {
          "url": "/project_stats/project_stats_list",
          "method": "get",
          "dataType": "json",
          "adaptor": function (payload) {
            return {
              status: payload.status,
              msg: payload.msg,
              data: {
                items: payload.data && payload.data.items ? payload.data.items : [],
                total: payload.data && payload.data.total ? payload.data.total : 0
              }
            };
          }
        },
        "headerToolbar": [
          {
            "type": "tpl",
            "tpl": "<div class='header-title'><i class='fa fa-cog'></i> 大直径盾构项目统计</div>"
          },
          "export-excel",
          {
            "type": "reload",
            "icon": "fa fa-refresh",
            "tooltip": "刷新数据",
            "tooltipPlacement": "top",
            "level": "link"
          }
        ],
        "filter": {
          "title": "查询条件",
          "submitText": "查询",
          "body": [
            {
              "type": "input-text",
              "name": "project_name",
              "label": "项目名称",
              "placeholder": "请输入项目名称进行查询"
            },
            {
              "type": "input-text",
              "name": "work_area",
              "label": "工区",
              "placeholder": "请输入工区进行查询"
            },
            {
              "type": "input-text",
              "name": "pro_code",
              "label": "项目编码",
              "placeholder": "请输入项目编码进行查询"
            }
          ]
        },
        "columns": [
          {
            "type": "operation",
            "label": "操作",
            "width": 80,
            "fixed": "right",
            "buttons": [
              {
                "type": "button",
                "icon": "fa fa-pencil",
                "tooltip": "修改",
                "level": "link",
                "actionType": "dialog",
                "dialog": {
                  "title": "修改项目统计",
                  "body": {
                    "type": "form",
                    "api": {
                      "method": "post",
                      "url": "/project_stats/update_project_stats",
                      "messages": {
                        "success": "更新成功",
                        "failed": "更新失败"
                      }
                    },
                    "body": [
                      {"type": "hidden", "name": "id"},
                      {"type": "input-text", "name": "project_name", "label": "项目名称", "required": true},
                      {"type": "input-text", "name": "work_area", "label": "工区", "required": true},
                      {"type": "input-text", "name": "bid_money", "label": "中标金额"},
                      {"type": "input-text", "name": "bid_time", "label": "中标时间"},
                      {"type": "input-text", "name": "finish_time", "label": "完工时间"},
                      {"type": "input-text", "name": "tunnel_length", "label": "隧道长度"},
                      {"type": "input-text", "name": "tube_outer_diameter", "label": "管片外径"},
                      {"type": "input-text", "name": "through_waters", "label": "是否穿越水域"},
                      {"type": "input-text", "name": "max_excavation_diameter", "label": "最大开挖直径"},
                      {"type": "input-text", "name": "project_type", "label": "工程类型"},
                      {"type": "input-text", "name": "job_schedule", "label": "工程进度"},
                      {"type": "input-text", "name": "type", "label": "后构类型"},
                      {"type": "input-text", "name": "company_name", "label": "单位名称"},
                      {"type": "input-text", "name": "province", "label": "省份"},
                      {"type": "input-text", "name": "remark", "label": "备注"},
                      {"type": "input-text", "name": "pro_code", "label": "项目编码"},
                      {"type": "input-text", "name": "sub_code", "label": "项目区间编码"},
                      {"type": "input-text", "name": "through_assise", "label": "穿越地层"},
                      {"type": "input-text", "name": "hydraulic_pressure", "label": "埋深水压"},
                      {"type": "input-text", "name": "intension", "label": "衬砌强度"},
                      {"type": "input-text", "name": "duct_piece", "label": "隧道管片"},
                      {"type": "input-number", "name": "zone_id", "label": "项目区间id"},
                      {"type": "input-number", "name": "shield_machine_id", "label": "盾构机ID"}
                    ]
                  }
                }
              },
              {
                "type": "button",
                "icon": "fa fa-times text-danger",
                "actionType": "ajax",
                "tooltip": "删除",
                "confirmText": "确定要删除该记录吗？",
                "api": {
                  "method": "delete",
                  "url": "/project_stats/delete_project_stats/${id}"
                },
                "messages": {
                  "success": "删除成功",
                  "failed": "删除失败"
                }
              }
            ]
          },
          { "name": "id","label": "编号","type": "number", "width": 80 },
          { "name": "project_name", "label": "项目名称", "type": "text", "width": 150 },
          { "name": "work_area", "label": "工区", "type": "text", "width": 100 },
          { "name": "bid_money", "label": "中标金额", "type": "text", "width": 120 },
          { "name": "bid_time", "label": "中标时间", "type": "text", "width": 120 },
          { "name": "finish_time", "label": "完工时间", "type": "text", "width": 120 },
          { "name": "tunnel_length", "label": "隧道长度", "type": "text", "width": 100 },
          { "name": "tube_outer_diameter", "label": "管片外径", "type": "text", "width": 100 },
          { "name": "through_waters", "label": "是否穿越水域", "type": "text", "width": 120 },
          { "name": "max_excavation_diameter", "label": "最大开挖直径", "type": "text", "width": 120 },
          { "name": "project_type", "label": "工程类型", "type": "text", "width": 100 },
          { "name": "job_schedule", "label": "工程进度", "type": "text", "width": 100 },
          { "name": "type", "label": "后构类型", "type": "text", "width": 100 },
          { "name": "company_name", "label": "单位名称", "type": "text", "width": 150 },
          { "name": "province", "label": "省份", "type": "text", "width": 100 },
          { "name": "remark", "label": "备注", "type": "text", "width": 200 },
          { "name": "pro_code", "label": "项目编码", "type": "text", "width": 120 },
          { "name": "sub_code", "label": "项目区间编码", "type": "text", "width": 120 },
          { "name": "through_assise", "label": "穿越地层", "type": "text", "width": 120 },
          { "name": "hydraulic_pressure", "label": "埋深水压", "type": "text", "width": 100 },
          { "name": "intension", "label": "衬砌强度", "type": "text", "width": 100 },
          { "name": "duct_piece", "label": "隧道管片", "type": "text", "width": 100 },
          { "name": "zone_id", "label": "项目区间id", "type": "number", "width": 100 },
          { "name": "shield_machine_id", "label": "盾构机ID", "type": "number", "width": 100 }
        ]
      }
    ]
  };

  window.jsonpCallback && window.jsonpCallback(response);
})();